lib/elixir/lib/calendar/date.ex

```diff
@@ -208,12 +207,6 @@ defmodule Date do
     range(first, last, step)
   end
 
-  def range(%{calendar: _, year: _, month: _, day: _} = first, duration, step) do
-    raise ArgumentError,
-          "expected a date or duration as second argument and the step must be a " <>
-            "non-zero integer, got: #{inspect(first)}, #{inspect(duration)}, #{step}"
-  end
```

Comment:
I think this was removed following this [comment](https://github.com/elixir-lang/elixir/pull/14172#discussion_r1912856557) from @whatyouhide, but a zero step won't be caught by the typesystem so I think we should probably still raise a proper `ArgumentError` in this case, WDYT?

Reply:
yes the zero step yep, good callout

Reply:
Looking at the [basic types](https://hexdocs.pm/elixir/1.12/typespecs.html#basic-types) an invalid step _should_ be caught by the type system (`pos_integer`, `neg_integer`), wouldn't it?

However, as it stands it's not consistent as `Date.range(d1, d2, 0)` raises an `ArgumentError` and  `Date.range(d1, duration, 0)` raises a `FunctionClauseError`.

I'll re-add the clauses with the adjusted ArgumentError message to `range/2` and `range/3`.

Reply:
also, how about not matching or guarding durations at all (as we also don't do it in shift/2)? so we'd just have:

```
def range(%{calendar: calendar} = first, %{calendar: calendar} = last)
def range(%{calendar: _, year: _, month: _, day: _}, %{calendar: _, year: _, month: _, day: _})
def range(%{calendar: _} = first, duration)
```

and

```
def range(%{calendar: calendar} = first, %{calendar: calendar} = last, step) when is_integer(step) and step != 0
def range(%{calendar: _, year: _, month: _, day: _}, %{calendar: _, year: _, month: _, day: _}, step)
def range(%{calendar: _} = first, duration) when is_integer(step) and step != 0
```

Reply:
> Looking at the [basic types](https://hexdocs.pm/elixir/1.12/typespecs.html#basic-types) an invalid step should be caught by the type system (pos_integer, neg_integer), wouldn't it?

These are the "old types" (typespecs), for now AFAIK there is no plan to implement these in the new set-theoretic types.

Reply:
oh actually wasn't aware of that distinction! thanks!